@damassi => add associated sale to the auction2 page

sweir27 opened the pull request:
Some auctions (usually benefit auctions) have multiple sales (such as an online-only pre sale + a live event). This PR adds the related sale component.

![image](https://cloud.githubusercontent.com/assets/2081340/24525552/b152c8ae-1568-11e7-862c-5af121c1418c.png)

Note! I am reusing your `upcomingLabel` util method, but it's a little weird that it lives under the `/current_auctions` component. Do you think we should add a `util` folder up a couple of levels? Or stick this file in `lib`? Open to ideas!

damassi commented:
> Note! I am reusing your upcomingLabel util method, but it's a little weird that it lives under the /current_auctions component. Do you think we should add a util folder up a couple of levels? Or stick this file in lib? Open to ideas!

I brought this up in #web a while back -- how it would be nice if we had a place to organize domain-specific util functions that can be shared across apps / components. @craigspaeth pointed me to the thought that with GraphQL, a good amount of that logic can now live inside of there, and is effectively lazy in the sense that you only request what you need so client apps remain slim while Metaphysics scales out horizontally. I'm not sure I'm totally sold on this because it seems a bit idealistic compared to something focused like `auctions/utils/*`, but that said I'm still relatively new to this new paradigm of thinking. 

I guess in short, I'm not sure!

damassi commented:
@sweir27 - One question -- since this is so similar to `<CurrentAuctions />` do you think it would be wise to drop this in there and perhaps share some code and style? Even if its different enough to warrant a new component, the styles should piggyback on each other I think. 

Thoughts?

sweir27 commented:
Updated this (component still looks good)
![image](https://cloud.githubusercontent.com/assets/2081340/24534187/4f035358-159a-11e7-9361-fc5fe917421f.png)

Now `CurrentAuction` renders a series of `AuctionBlock`s, but you can also mount an `AuctionBlock` by itself, as we do for the related auction on the auction2 page.

One thought is that I'm currently passing all of the individual fields necessary to render an `AuctionBlock` as props, while I could be passing a `sale` object. I'm not sure what's best here. On the one hand, it's nice to explicitly require fields, but on the other hand it's a little annoying to keep repeating them. I'm open to either option, but at least this way the component is self-explanatory.

Also I added a spec for `upcoming_label` and switched around the logic a bit so it _always_ respects the `preview` and `closed` states.

sweir27 commented on desktop/apps/auction2/client.js:
```diff
@@ -44,6 +45,15 @@ const calendar = new AddToCalendar({
+  const associatedSale = auction.get('associated_sale')
+  mountAuctionBlock({
+    relatedAuction: true,
+    sale: associatedSale
```
Updated this to accept a `sale` instead of all of the individual attributes as props!

damassi commented on desktop/components/auction_block/index.jsx:
```diff
@@ -0,0 +1,25 @@
+import React from 'react'
+import ReactDOM from 'react-dom'
+import { isEmpty, isString } from 'underscore'
+
```
One thing I would do here to make it a bit more API like is

```javascript
export {
   AuctionBlock
}
```

so that later, rather than having to import the full path from within the current auctions component, we can just do 

```javascript
import { AuctionBlock } from 'components/auction_block'`
...
```

sweir27 replied:
Whoa cool! Didn't know you could do that.

damassi commented on desktop/components/auction_block/index.jsx:
```diff
@@ -0,0 +1,25 @@
+import ReactDOM from 'react-dom'
+import { isEmpty, isString } from 'underscore'
+
+export default function mount (data, selector = '#react-mount-current-auctions') {
```
Update selector to new default `#react-mount-auction-block`

damassi commented on desktop/components/auction_block/index.jsx:
```diff
@@ -0,0 +1,25 @@
+  } = data
+
+  const shouldMount = !isEmpty(sale.id) && isString(selector)
+
```
This is something I've been meaning to do in `current_auctions` but haven't gotten around to it yet (do you think you could update for me?) -- add in a runtime invariant (which just passes right through if truthy, but warns if false) that warns the user if something is up with their props:

```javascript
import invariant from 'invariant' 
... 
const shouldMount = !isEmpty(sale.id) && isString(selector)

invariant(shouldMount, 
  `Error mounting <AuctionBlock />: sale.id (${sale.id}) or `selector` (${selector}) is invalid.`
)
````
Or something like the above; could even move it into an else at the bottom -- something to alert devs that something is wrong with the operation / `shouldMount` is false.)

damassi commented on desktop/components/current_auctions/current_auctions.jsx:
```diff
@@ -1,6 +1,6 @@
 import React, { PropTypes } from 'react'
-import upcomingLabel from './utils/upcoming_label'
 import { partition, take } from 'underscore'
+import AuctionBlock from '../auction_block/auction_block.jsx'
```
Update with simpler format as described above

damassi replied:
Oh one thing -- could you update this?

damassi commented on desktop/apps/auction2/queries/sale.js:
```diff
@@ -2,6 +2,22 @@ export default function SaleQuery(id) {
+        live_start_at
+        name
+        start_at
+      }
```
Isn't the standard practice to co-locate a sufficiently renderable metaphysics query alongside the `<Component />` even if its not directly used within the code? So that when using it elsewhere you can import and compose it as a string similar to what we did with `current_auctions` in `artwork/components/client.coffee`?

damassi commented:
LGTM, will merge on 📗
